File: Task_2/task_2.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static string FindFinalDestination(List<(string, string)> routes)
    {
        var graph = new Dictionary<string, string>();
        var destinations = new HashSet<string>();

        foreach (var (start, end) in routes)
        {
            graph[start] = end;
            destinations.Add(end);
        }

        var starts = routes.Select(r => r.Item1).ToHashSet().Except(destinations);
        if (!starts.Any()) return "обратитесь к специалисту";

        string finalDestination = "";
        int maxLength = 0;

        foreach (var start in starts)
        {
            var current = start;
            int length = 1;

            while (graph.ContainsKey(current))
            {
                current = graph[current];
                length++;
            }

            if (length > maxLength)
            {
                maxLength = length;
                finalDestination = current;
            }
        }

        if (maxLength == 1 && finalDestination == graph.GetValueOrDefault(finalDestination))
        {
            return "обратитесь к специалисту";
        }

        return finalDestination;
    }

    static void Main()
    {
        var routes = new List<(string, string)>();
        Console.WriteLine("Введите пары маршрутов в формате 'начальный пункт конечный пункт'.");
        Console.WriteLine("Для завершения ввода введите пустую строку.");

        while (true)
        {
            Console.Write("Введите маршрут: ");
            var input = Console.ReadLine();
            if (string.IsNullOrEmpty(input)) break;

            var parts = input.Split(' ');
            if (parts.Length == 2)
            {
                routes.Add((parts[0], parts[1]));
            }
            else
            {
                Console.WriteLine("Некорректный ввод. Попробуйте снова.");
            }
        }

        var result = FindFinalDestination(routes);
        Console.WriteLine($"\n{result}");
    }
}

File: Task_2/task_2.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"
)

func findFinalDestination(routes [][2]string) string {
    graph := make(map[string]string)
    destinations := make(map[string]bool)

    for _, route := range routes {
        start, end := route[0], route[1]
        graph[start] = end
        destinations[end] = true
    }

    starts := []string{}
    for _, route := range routes {
        start := route[0]
        if !destinations[start] {
            starts = append(starts, start)
        }
    }

    if len(starts) == 0 {
        return "обратитесь к специалисту"
    }

    var finalDestination string
    maxLength := 0

    for _, start := range starts {
        current := start
        length := 1

        for {
            next, ok := graph[current]
            if !ok {
                break
            }
            current = next
            length++
        }

        if length > maxLength {
            maxLength = length
            finalDestination = current
        }
    }

    if maxLength == 1 && finalDestination == graph[finalDestination] {
        return "обратитесь к специалисту"
    }

    return finalDestination
}

func main() {
    routes := [][2]string{}
    scanner := bufio.NewScanner(os.Stdin)

    fmt.Println("Введите пары маршрутов в формате 'начальный пункт конечный пункт'.")
    fmt.Println("Для завершения ввода введите пустую строку.")

    for {
        fmt.Print("Введите маршрут: ")
        scanner.Scan()
        input := strings.TrimSpace(scanner.Text())
        if input == "" {
            break
        }

        parts := strings.Split(input, " ")
        if len(parts) == 2 {
            routes = append(routes, [2]string{parts[0], parts[1]})
        } else {
            fmt.Println("Некорректный ввод. Попробуйте снова.")
        }
    }

    result := findFinalDestination(routes)
    fmt.Printf("\n%s\n", result)
}